google/cloud/logging_v2/_instrumentation.py

```diff
@@ -0,0 +1,156 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Add diagnostic instrumentation source information to logs"""
+
+from google.cloud.logging_v2.entries import StructEntry
+from google.cloud.logging_v2 import __version__
+
+_DIAGNOSTIC_INFO_KEY = "logging.googleapis.com/diagnostic"
+_INSTRUMENTATION_SOURCE_KEY = "instrumentation_source"
+_PYTHON_LIBRARY_NAME = "python"
+
+_LIBRARY_VERSION = __version__
+
+_MAX_NAME_LENGTH = 14
+_MAX_VERSION_LENGTH = 16
+
+
+def add_instrumentation(entries, **kw):
+    """Validate or add instrumentation information to a list of entries
+
+    Args:
+       entries (Sequence[Mapping[str, ...]]): sequence of mappings representing
+            the log entry resources to log.
+
+    Returns:
+        Sequence[Mapping[str, ...]]: entries with instrumentation info validated if present
+        otherwise added to beginning of list.
+    """
+    is_written = False
+    for entry in entries:
+        if (
+            is_written is False
+            and hasattr(entry, "payload")
+            and entry.payload is dict
+            and _DIAGNOSTIC_INFO_KEY in entry.payload
+            and _INSTRUMENTATION_SOURCE_KEY in entry.payload[_DIAGNOSTIC_INFO_KEY]
+        ):
+            current_info = entry.payload[_DIAGNOSTIC_INFO_KEY][
+                _INSTRUMENTATION_SOURCE_KEY
+            ]
+            entry.payload[_DIAGNOSTIC_INFO_KEY][
+                _INSTRUMENTATION_SOURCE_KEY
+            ] = validate_and_update_instrumentation(current_info)
+            is_written = True
+        else:
+            diagnostic_entry = create_diagnostic_entry(**kw)
+            entries.insert(0, diagnostic_entry.to_api_repr())
+        return entries
+
+
+def create_diagnostic_entry(name=None, version=None, **kw):
+    """Create a diagnostic log entry describing this library
+
+    Args:
+        name(str): The name of this library (e.g. 'python')
+        version(str) The version of this library (e.g. '3.0.0')
+
+    Returns:
+        google.cloud.logging_v2.LogEntry: Log entry with library information
+    """
+    payload = {
+        _DIAGNOSTIC_INFO_KEY: {
+            _INSTRUMENTATION_SOURCE_KEY: [_get_instrumentation_source(name, version)]
+        }
+    }
+    kw["severity"] = "INFO"
+    entry = StructEntry(payload=payload, **kw)
+    return entry
+
+
+def truncate_string(str, max_length):
+    """Truncate a string to a maximum length
+
+    Args:
+        str(str): The string to truncate
+        max_length(int): The maximum length
+
+    Returns:
+        A string containing either 'str' or a truncated version of
+        'str' with an asterisk at the end
+    """
+    if len(str) > max_length:
+        return str[: max_length - 1] + "*"
+    else:
+        return str
+
+
+def validate_and_update_instrumentation(existing_info=None):
+    """Validate existing instrumentation info and append the final entry
```

Comment:
It only matters if we ever expect to receive a log entry with existing (or partial) instrumentation source.  This is probably not going to happen with the library as it stands, but in the future, something could be built on top of this library (like the bunyan or winston integrations for nodejs) that adds a valid 'layer' of instrumentation info that we don't want to clobber.

It would certainly simplify things to remove this logic and just always emit a single entry with just this library's info... What do you think?